File: ios/App/App/SecureVaultPlugin.swift
import Foundation
import Capacitor
import CryptoKit
import LocalAuthentication
import Security

@objc(SecureVaultPlugin)
public class SecureVaultPlugin: CAPPlugin {
    private let service = "stream.wallet.vault.v1"
    private let account = "encryption-key"
    private let payloadKey = "stream.wallet.vault.ciphertext"
    private let addressKey = "stream.wallet.vault.address"

    @objc public func status(_ call: CAPPluginCall) {
        let context = LAContext()
        var error: NSError?
        let available = context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error)
        let defaults = UserDefaults.standard
        let exists = defaults.string(forKey: payloadKey) != nil
            && defaults.string(forKey: addressKey) != nil
            && keyExists()

        var result: [String: Any] = [
            "available": available,
            "exists": exists,
            "platform": "ios",
            "protection": "biometric-or-device-passcode"
        ]
        if exists, let address = defaults.string(forKey: addressKey) {
            result["address"] = address
        }
        call.resolve(result)
    }

    @objc public func save(_ call: CAPPluginCall) {
        guard let rawMnemonic = call.getString("mnemonic") else {
            call.reject("A validated 12-word recovery phrase is required.", "invalid_mnemonic")
            return
        }
        let mnemonic = rawMnemonic
            .split(whereSeparator: { $0.isWhitespace })
            .map(String.init)
            .joined(separator: " ")
        guard mnemonic.split(separator: " ").count == 12 else {
            call.reject("A validated 12-word recovery phrase is required.", "invalid_mnemonic")
            return
        }
        guard let address = call.getString("address"), address.hasPrefix("zkas:") else {
            call.reject("A validated ZKAS mainnet address is required.", "invalid_address")
            return
        }
        guard let accountSeedHex = call.getString("accountSeedHex"), isHex(accountSeedHex, count: 64),
              let fvkHex = call.getString("fvkHex"), isHex(fvkHex, count: 192),
              let walletToken = call.getString("walletToken"), isHex(walletToken, count: 32) else {
            call.reject("Validated mainnet wallet material is required.", "invalid_wallet_material")
            return
        }
        let birthdayDaa = max(0, call.getInt("birthdayDaa") ?? 0)
        guard UserDefaults.standard.string(forKey: payloadKey) == nil && !keyExists() else {
            call.reject("A wallet already exists on this device.", "vault_exists")
            return
        }

        authenticate(reason: "Protect your Stream Wallet recovery phrase") { [weak self] success, authError in
            guard let self = self else { return }
            guard success else {
                call.reject(authError?.localizedDescription ?? "Device authentication was cancelled.", "authentication_cancelled")
                return
            }
            DispatchQueue.global(qos: .userInitiated).async {
                var keyData = Data(count: 32)
                let randomStatus = keyData.withUnsafeMutableBytes { bytes in
                    SecRandomCopyBytes(kSecRandomDefault, 32, bytes.baseAddress!)
                }
                guard randomStatus == errSecSuccess else {
                    self.rejectOnMain(call, "Secure random key generation failed.", "secure_storage_unavailable")
                    return
                }

                do {
                    try self.storeKey(keyData)
                    let key = SymmetricKey(data: keyData)
                    var plaintext = try JSONSerialization.data(withJSONObject: [
                        "version": 2,
                        "mnemonic": mnemonic,
                        "address": address,
                        "accountSeedHex": accountSeedHex.lowercased(),
                        "fvkHex": fvkHex.lowercased(),
                        "walletToken": walletToken.lowercased(),
                        "birthdayDaa": birthdayDaa
                    ])
                    defer { plaintext.resetBytes(in: 0..<plaintext.count) }
                    let sealed = try AES.GCM.seal(plaintext, using: key)
                    guard let combined = sealed.combined else {
                        throw VaultError.invalidCiphertext
                    }
                    UserDefaults.standard.set(combined.base64EncodedString(), forKey: self.payloadKey)
                    UserDefaults.standard.set(address, forKey: self.addressKey)
                    keyData.resetBytes(in: 0..<keyData.count)
                    self.resolveOnMain(call, ["address": address])
                } catch {
                    keyData.resetBytes(in: 0..<keyData.count)
                    self.deleteKey()
                    UserDefaults.standard.removeObject(forKey: self.payloadKey)
                    UserDefaults.standard.removeObject(forKey: self.addressKey)
                    self.rejectOnMain(call, "The encrypted vault could not be saved.", "vault_save_failed", error)
                }
            }
        }
    }

    @objc public func unlock(_ call: CAPPluginCall) {
        openVault(call, removeAfter: false, authorizeSpend: false)
    }

    @objc public func authorize(_ call: CAPPluginCall) {
        openVault(call, removeAfter: false, authorizeSpend: true)
    }

    @objc public func remove(_ call: CAPPluginCall) {
        openVault(call, removeAfter: true, authorizeSpend: false)
    }

    private func openVault(_ call: CAPPluginCall, removeAfter: Bool, authorizeSpend: Bool) {
        let defaults = UserDefaults.standard
        guard let encoded = defaults.string(forKey: payloadKey),
              let combined = Data(base64Encoded: encoded),
              let address = defaults.string(forKey: addressKey),
              keyExists() else {
            call.reject("No complete wallet vault exists on this device.", "vault_missing")
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let reason = removeAfter
                    ? "Confirm permanent removal of Stream Wallet"
                    : authorizeSpend ? "Authorize this ZKAS mainnet payment" : "Unlock Stream Wallet"
                var keyData = try self.loadKey(reason: reason)
                defer { keyData.resetBytes(in: 0..<keyData.count) }
                let key = SymmetricKey(data: keyData)
                let box = try AES.GCM.SealedBox(combined: combined)
                var plaintext = try AES.GCM.open(box, using: key)
                defer { plaintext.resetBytes(in: 0..<plaintext.count) }
                if let legacyMnemonic = String(data: plaintext, encoding: .utf8),
                   !legacyMnemonic.trimmingCharacters(in: .whitespacesAndNewlines).hasPrefix("{"),
                   legacyMnemonic.split(whereSeparator: { $0.isWhitespace }).count == 12 {
                    if removeAfter {
                        defaults.removeObject(forKey: self.payloadKey)
                        defaults.removeObject(forKey: self.addressKey)
                        self.deleteKey()
                        self.resolveOnMain(call, ["removed": true])
                        return
                    }
                    throw VaultError.upgradeRequired
                }
                guard let payload = try JSONSerialization.jsonObject(with: plaintext) as? [String: Any],
                      let mnemonic = payload["mnemonic"] as? String,
                      mnemonic.split(whereSeparator: { $0.isWhitespace }).count == 12,
                      let payloadAddress = payload["address"] as? String,
                      payloadAddress == address,
                      let accountSeedHex = payload["accountSeedHex"] as? String,
                      self.isHex(accountSeedHex, count: 64),
                      let fvkHex = payload["fvkHex"] as? String,
                      self.isHex(fvkHex, count: 192),
                      let walletToken = payload["walletToken"] as? String,
                      self.isHex(walletToken, count: 32) else {
                    throw VaultError.invalidPlaintext
                }
                let birthdayDaa = payload["birthdayDaa"] as? Int ?? 0

                if removeAfter {
                    defaults.removeObject(forKey: self.payloadKey)
                    defaults.removeObject(forKey: self.addressKey)
                    self.deleteKey()
                    self.resolveOnMain(call, ["removed": true])
                } else if authorizeSpend {
                    self.resolveOnMain(call, [
                        "address": address,
                        "accountSeedHex": accountSeedHex
                    ])
                } else {
                    self.resolveOnMain(call, [
                        "address": address,
                        "fvkHex": fvkHex,
                        "walletToken": walletToken,
                        "birthdayDaa": birthdayDaa
                    ])
                }
            } catch VaultError.authenticationFailed(let status) {
                let error = NSError(domain: NSOSStatusErrorDomain, code: Int(status))
                self.rejectOnMain(call, "Device authentication was cancelled or failed.", "authentication_cancelled", error)
            } catch let error as NSError where error.domain == LAError.errorDomain {
                self.rejectOnMain(call, error.localizedDescription, "authentication_cancelled", error)
            } catch VaultError.upgradeRequired {
                self.rejectOnMain(call, "This pre-mainnet vault must be removed and restored from its recovery phrase.", "vault_upgrade_required")
            } catch {
                self.rejectOnMain(call,
                    removeAfter ? "The wallet could not be removed." : "The wallet could not be unlocked.",
                    removeAfter ? "vault_remove_failed" : "vault_unlock_failed",
                    error)
            }
        }
    }

    private func authenticate(reason: String, completion: @escaping (Bool, Error?) -> Void) {
        let context = LAContext()
        context.localizedCancelTitle = "Cancel"
        context.evaluatePolicy(.deviceOwnerAuthentication, localizedReason: reason) { success, error in
            DispatchQueue.main.async { completion(success, error) }
        }
    }

    private func storeKey(_ key: Data) throws {
        var accessError: Unmanaged<CFError>?
        guard let access = SecAccessControlCreateWithFlags(
            nil,
            kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
            .userPresence,
            &accessError
        ) else {
            throw accessError?.takeRetainedValue() ?? VaultError.keychainFailure(errSecParam)
        }

        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecAttrSynchronizable as String: false,
            kSecAttrAccessControl as String: access,
            kSecValueData as String: key
        ]
        let status = SecItemAdd(query as CFDictionary, nil)
        guard status == errSecSuccess else { throw VaultError.keychainFailure(status) }
    }

    private func loadKey(reason: String) throws -> Data {
        let context = LAContext()
        context.localizedReason = reason
        context.localizedCancelTitle = "Cancel"
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecMatchLimit as String: kSecMatchLimitOne,
            kSecReturnData as String: true,
            kSecUseAuthenticationContext as String: context
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess, let data = item as? Data else {
            if status == errSecUserCanceled || status == errSecAuthFailed || status == errSecInteractionNotAllowed {
                throw VaultError.authenticationFailed(status)
            }
            throw VaultError.keychainFailure(status)
        }
        return data
    }

    private func keyExists() -> Bool {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecMatchLimit as String: kSecMatchLimitOne,
            kSecReturnAttributes as String: true,
            kSecUseAuthenticationUI as String: kSecUseAuthenticationUIFail
        ]
        let status = SecItemCopyMatching(query as CFDictionary, nil)
        return status == errSecSuccess || status == errSecInteractionNotAllowed
    }

    private func deleteKey() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account
        ]
        SecItemDelete(query as CFDictionary)
    }

    private func isHex(_ value: String, count: Int) -> Bool {
        guard value.count == count else { return false }
        return value.allSatisfy { $0.isHexDigit }
    }

    private func resolveOnMain(_ call: CAPPluginCall, _ value: [String: Any]) {
        DispatchQueue.main.async { call.resolve(value) }
    }

    private func rejectOnMain(_ call: CAPPluginCall, _ message: String, _ code: String, _ error: Error? = nil) {
        DispatchQueue.main.async { call.reject(message, code, error) }
    }
}

private enum VaultError: Error {
    case invalidCiphertext
    case invalidPlaintext
    case authenticationFailed(OSStatus)
    case keychainFailure(OSStatus)
    case upgradeRequired
}

File: android/app/src/main/java/stream/zkas/wallet/security/SecureVaultPlugin.java
package stream.zkas.wallet.security;

import android.content.SharedPreferences;
import android.os.Build;
import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyPermanentlyInvalidatedException;
import android.security.keystore.KeyProperties;
import android.util.Base64;

import androidx.annotation.NonNull;
import androidx.biometric.BiometricManager;
import androidx.biometric.BiometricPrompt;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.nio.charset.StandardCharsets;
import java.security.KeyStore;
import java.util.Arrays;
import java.util.concurrent.Executor;

import org.json.JSONObject;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;

@CapacitorPlugin(name = "SecureVault")
public class SecureVaultPlugin extends Plugin {
    private static final String KEY_ALIAS = "stream.wallet.vault.v1";
    private static final String PREFS = "stream_wallet_vault";
    private static final String FIELD_CIPHERTEXT = "ciphertext";
    private static final String FIELD_IV = "iv";
    private static final String FIELD_ADDRESS = "address";
    private static final String CIPHER = "AES/GCM/NoPadding";

    @PluginMethod
    public void status(PluginCall call) {
        SharedPreferences prefs = prefs();
        boolean exists = prefs.contains(FIELD_CIPHERTEXT)
                && prefs.contains(FIELD_IV)
                && prefs.contains(FIELD_ADDRESS)
                && keyExists();
        int capability = BiometricManager.from(getContext()).canAuthenticate(authenticators());

        JSObject result = new JSObject();
        result.put("available", capability == BiometricManager.BIOMETRIC_SUCCESS);
        result.put("exists", exists);
        result.put("platform", "android");
        result.put("protection", Build.VERSION.SDK_INT >= Build.VERSION_CODES.R
                ? "biometric-or-device-credential"
                : "strong-biometric");
        if (exists) result.put("address", prefs.getString(FIELD_ADDRESS, ""));
        call.resolve(result);
    }

    @PluginMethod
    public void save(PluginCall call) {
        String mnemonic = call.getString("mnemonic");
        String address = call.getString("address");
        String accountSeedHex = call.getString("accountSeedHex");
        String fvkHex = call.getString("fvkHex");
        String walletToken = call.getString("walletToken");
        int birthdayDaa = Math.max(0, call.getInt("birthdayDaa", 0));
        if (mnemonic == null || mnemonic.trim().split("\\s+").length != 12) {
            call.reject("A validated 12-word recovery phrase is required.", "invalid_mnemonic");
            return;
        }
        if (address == null || !address.startsWith("zkas:")) {
            call.reject("A validated ZKAS mainnet address is required.", "invalid_address");
            return;
        }
        if (!isHex(accountSeedHex, 64) || !isHex(fvkHex, 192) || !isHex(walletToken, 32)) {
            call.reject("Validated mainnet wallet material is required.", "invalid_wallet_material");
            return;
        }
        if (prefs().contains(FIELD_CIPHERTEXT) || keyExists()) {
            call.reject("A wallet already exists on this device.", "vault_exists");
            return;
        }

        try {
            SecretKey key = generateKey();
            Cipher cipher = Cipher.getInstance(CIPHER);
            cipher.init(Cipher.ENCRYPT_MODE, key);
            authenticate(call, cipher, "Protect Stream Wallet", "Confirm to encrypt this recovery phrase", authenticated -> {
                byte[] plaintext = null;
                try {
                    JSONObject payload = new JSONObject();
                    payload.put("version", 2);
                    payload.put("mnemonic", mnemonic.trim().replaceAll("\\s+", " "));
                    payload.put("address", address);
                    payload.put("accountSeedHex", accountSeedHex.toLowerCase());
                    payload.put("fvkHex", fvkHex.toLowerCase());
                    payload.put("walletToken", walletToken.toLowerCase());
                    payload.put("birthdayDaa", birthdayDaa);
                    plaintext = payload.toString().getBytes(StandardCharsets.UTF_8);
                    byte[] encrypted = authenticated.doFinal(plaintext);
                    boolean saved = prefs().edit()
                            .putString(FIELD_CIPHERTEXT, Base64.encodeToString(encrypted, Base64.NO_WRAP))
                            .putString(FIELD_IV, Base64.encodeToString(authenticated.getIV(), Base64.NO_WRAP))
                            .putString(FIELD_ADDRESS, address)
                            .commit();
                    Arrays.fill(encrypted, (byte) 0);
                    if (!saved) throw new IllegalStateException("Encrypted vault could not be committed.");
                    JSObject result = new JSObject();
                    result.put("address", address);
                    call.resolve(result);
                } catch (Exception error) {
                    deleteKey();
                    prefs().edit().clear().commit();
                    call.reject("The encrypted vault could not be saved.", "vault_save_failed", error);
                } finally {
                    if (plaintext != null) Arrays.fill(plaintext, (byte) 0);
                }
            }, this::deleteKey);
        } catch (Exception error) {
            deleteKey();
            call.reject("Secure device storage could not be initialized.", "secure_storage_unavailable", error);
        }
    }

    @PluginMethod
    public void unlock(PluginCall call) {
        decryptAfterAuthentication(call, "Unlock Stream Wallet", "Confirm to open this wallet", false, false);
    }

    @PluginMethod
    public void authorize(PluginCall call) {
        decryptAfterAuthentication(call, "Authorize ZKAS payment", "Confirm this mainnet transaction", false, true);
    }

    @PluginMethod
    public void remove(PluginCall call) {
        decryptAfterAuthentication(call, "Remove Stream Wallet", "Confirm permanent removal from this device", true, false);
    }

    private void decryptAfterAuthentication(PluginCall call, String title, String subtitle, boolean removeAfter, boolean authorizeSpend) {
        SharedPreferences prefs = prefs();
        String ciphertext = prefs.getString(FIELD_CIPHERTEXT, null);
        String iv = prefs.getString(FIELD_IV, null);
        String address = prefs.getString(FIELD_ADDRESS, null);
        if (ciphertext == null || iv == null || address == null || !keyExists()) {
            call.reject("No complete wallet vault exists on this device.", "vault_missing");
            return;
        }

        try {
            SecretKey key = loadKey();
            Cipher cipher = Cipher.getInstance(CIPHER);
            cipher.init(Cipher.DECRYPT_MODE, key,
                    new GCMParameterSpec(128, Base64.decode(iv, Base64.NO_WRAP)));
            authenticate(call, cipher, title, subtitle, authenticated -> {
                byte[] encrypted = Base64.decode(ciphertext, Base64.NO_WRAP);
                byte[] plaintext = null;
                try {
                    plaintext = authenticated.doFinal(encrypted);
                    String decoded = new String(plaintext, StandardCharsets.UTF_8);
                    if (decoded.trim().split("\\s+").length == 12 && !decoded.trim().startsWith("{")) {
                        if (removeAfter) {
                            if (!prefs.edit().clear().commit()) {
                                throw new IllegalStateException("Vault data could not be removed.");
                            }
                            deleteKey();
                            JSObject removed = new JSObject();
                            removed.put("removed", true);
                            call.resolve(removed);
                        } else {
                            call.reject("This pre-mainnet vault must be removed and restored from its recovery phrase.",
                                    "vault_upgrade_required");
                        }
                        return;
                    }
                    JSONObject payload = new JSONObject(decoded);
                    String mnemonic = payload.optString("mnemonic", "");
                    String payloadAddress = payload.optString("address", "");
                    String accountSeedHex = payload.optString("accountSeedHex", "");
                    String fvkHex = payload.optString("fvkHex", "");
                    String walletToken = payload.optString("walletToken", "");
                    int birthdayDaa = Math.max(0, payload.optInt("birthdayDaa", 0));
                    if (mnemonic.trim().split("\\s+").length != 12
                            || !address.equals(payloadAddress)
                            || !isHex(accountSeedHex, 64)
                            || !isHex(fvkHex, 192)
                            || !isHex(walletToken, 32)) {
                        throw new IllegalStateException("Decrypted vault contents are invalid.");
                    }
                    JSObject result = new JSObject();
                    if (removeAfter) {
                        if (!prefs.edit().clear().commit()) {
                            throw new IllegalStateException("Vault data could not be removed.");
                        }
                        deleteKey();
                        result.put("removed", true);
                    } else if (authorizeSpend) {
                        result.put("address", address);
                        result.put("accountSeedHex", accountSeedHex);
                    } else {
                        result.put("address", address);
                        result.put("fvkHex", fvkHex);
                        result.put("walletToken", walletToken);
                        result.put("birthdayDaa", birthdayDaa);
                    }
                    call.resolve(result);
                } catch (Exception error) {
                    call.reject(removeAfter ? "The wallet could not be removed." : "The wallet could not be unlocked.",
                            removeAfter ? "vault_remove_failed" : "vault_unlock_failed", error);
                } finally {
                    Arrays.fill(encrypted, (byte) 0);
                    if (plaintext != null) Arrays.fill(plaintext, (byte) 0);
                }
            }, null);
        } catch (KeyPermanentlyInvalidatedException error) {
            call.reject("Device security changed and invalidated the vault key. Restore from the recovery phrase.",
                    "vault_key_invalidated", error);
        } catch (Exception error) {
            call.reject("The secure vault could not be opened.", "vault_unlock_failed", error);
        }
    }

    private void authenticate(
            PluginCall call,
            Cipher cipher,
            String title,
            String subtitle,
            CipherConsumer onSuccess,
            Runnable onFailure
    ) {
        FragmentActivity activity = (FragmentActivity) getActivity();
        Executor executor = ContextCompat.getMainExecutor(getContext());
        BiometricPrompt prompt = new BiometricPrompt(activity, executor, new BiometricPrompt.AuthenticationCallback() {
            @Override
            public void onAuthenticationError(int errorCode, @NonNull CharSequence errString) {
                if (onFailure != null) onFailure.run();
                call.reject(errString.toString(), "authentication_cancelled");
            }

            @Override
            public void onAuthenticationSucceeded(@NonNull BiometricPrompt.AuthenticationResult result) {
                Cipher authenticated = result.getCryptoObject() == null ? null : result.getCryptoObject().getCipher();
                if (authenticated == null) {
                    if (onFailure != null) onFailure.run();
                    call.reject("Device authentication returned no cryptographic authorization.", "authentication_failed");
                    return;
                }
                onSuccess.accept(authenticated);
            }

            @Override
            public void onAuthenticationFailed() {
                // The system prompt stays open and allows another attempt.
            }
        });

        BiometricPrompt.PromptInfo.Builder info = new BiometricPrompt.PromptInfo.Builder()
                .setTitle(title)
                .setSubtitle(subtitle)
                .setConfirmationRequired(true)
                .setAllowedAuthenticators(authenticators());
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) info.setNegativeButtonText("Cancel");
        prompt.authenticate(info.build(), new BiometricPrompt.CryptoObject(cipher));
    }

    private SecretKey generateKey() throws Exception {
        KeyGenerator generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore");
        KeyGenParameterSpec.Builder spec = new KeyGenParameterSpec.Builder(
                KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
                .setKeySize(256)
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setRandomizedEncryptionRequired(true)
                .setUserAuthenticationRequired(true);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            spec.setUserAuthenticationParameters(0,
                    KeyProperties.AUTH_BIOMETRIC_STRONG | KeyProperties.AUTH_DEVICE_CREDENTIAL);
        } else {
            spec.setUserAuthenticationValidityDurationSeconds(-1);
        }
        generator.init(spec.build());
        return generator.generateKey();
    }

    private SecretKey loadKey() throws Exception {
        KeyStore store = KeyStore.getInstance("AndroidKeyStore");
        store.load(null);
        return (SecretKey) store.getKey(KEY_ALIAS, null);
    }

    private boolean keyExists() {
        try {
            KeyStore store = KeyStore.getInstance("AndroidKeyStore");
            store.load(null);
            return store.containsAlias(KEY_ALIAS);
        } catch (Exception ignored) {
            return false;
        }
    }

    private void deleteKey() {
        try {
            KeyStore store = KeyStore.getInstance("AndroidKeyStore");
            store.load(null);
            if (store.containsAlias(KEY_ALIAS)) store.deleteEntry(KEY_ALIAS);
        } catch (Exception ignored) {
            // A failed cleanup is surfaced the next time status/save checks the alias.
        }
    }

    private int authenticators() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            return BiometricManager.Authenticators.BIOMETRIC_STRONG
                    | BiometricManager.Authenticators.DEVICE_CREDENTIAL;
        }
        return BiometricManager.Authenticators.BIOMETRIC_STRONG;
    }

    private SharedPreferences prefs() {
        return getContext().getSharedPreferences(PREFS, 0);
    }

    private boolean isHex(String value, int length) {
        return value != null && value.length() == length && value.matches("[0-9a-fA-F]+$");
    }

    private interface CipherConsumer {
        void accept(Cipher cipher);
    }
}
